Add _FillValue attribute to most 3D fields written to output files

wenshanw opened the pull request:
This merge 
1. adds the `_FillValue` attribute to all fields that get written to `timeSeriesStatsMonthlyOutput` by default
2. replaces `-1e34`  and other arbitrary invalid values with the NetCDF `_FillValue`.

Variables where `_FillValue` could not be added are:
* Invalid values are zeros instead of `_FillValue`
  - `var_array activeTracersTend`
* Causing errors in case runs
  - `var wettingVelocity`
  - `var_array name="vertNonLocalFlux`

wenshanw commented:
Problem remained: if I add fill_value to `wettingVelocity` or `var_array name="vertNonLocalFlux"`, both the test and daily average cases crashed at the running time. Not much info from the error msg. Adding fill_value to all other variables is fine. Do you guys have any ideas?

xylar reviewed (commented):
@wenshanw, were you able to run the QU240/daily_output_test` with these changes?  I'm seeing an error:
```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x7f3da239dfcf in ???
#1  0x564692971d21 in __ocn_time_integration_split_MOD_ocn_time_integrator_split
	at /home/xylar/code/mpas-work/model/ocean/test_merge/src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:967
#2  0x5646929820b3 in __ocn_time_integration_MOD_ocn_timestep
	at /home/xylar/code/mpas-work/model/ocean/test_merge/src/core_ocean/mode_forward/mpas_ocn_time_integration.F:111
...
```

philipwjones commented:
@xylar - I have been seeing ieee underflow and inexact errors on Summit for QU240 with the latest ocean/develop (run still completes since we're not trapping those by default in optimized mode). Was starting to track it down yesterday but haven't gotten far.

xylar commented:
@philipwjones, I've been seeing those, too, approximately since the merge of `ocean/coastal` to `ocean/develop` (though that may be a red herring).  I'm seeing them with gnu but not with intel.  I don't think they are related to this work.  I see those warnings with `ocean/develop` as well.

Very glad you're looking into those because they had me worried.

xylar reviewed (commented):
I added all of the `missing_value="FILLVAL"` into the main registry, one `var_struct` at a time.  Everything worked for me.  I didn't add fill values to fields in the `scratch` `var_struct` because we would never be outputting fields from that.  (I certainly hope we aren't!)

I got the feeling that fill values were added to a lot of variables that aren't in `timeSeriesStatsDaily/Monthly`.  That's okay but I wanted to check on that just to understand what your system was for deciding which variables got fill values and which didn't.


In my testing, the following variables in `timeSeriesStatsDaily` ended up with `_FillValue` attributes but still have `0` as the missing value in the output:
* `timeDaily_avg_velocityMeridional`
* `timeDaily_avg_velocityZonal`
* `timeDaily_avg_tendLayerThickness`
* `timeDaily_avg_normalGMBolusVelocity`
* `timeDaily_avg_GMBolusVelocityZonal`
* `timeDaily_avg_GMBolusVelocityMeridional`
* `timeDaily_avg_vertDiffTopOfCell`
* `timeDaily_avg_vertViscTopOfCell`
* `timeDaily_avg_bulkRichardsonNumber` (?)
* `timeDaily_avg_tendLayerThickness`
* `timeDaily_avg_activeTracerVertMixTendency_temperatureVertMixTendency`
* `timeDaily_avg_activeTracerVertMixTendency_salinityVertMixTendency`
* `timeDaily_avg_activeTracerSurfaceFluxTendency_temperatureSurfaceFluxTendency`
* `timeDaily_avg_activeTracerSurfaceFluxTendency_salinitySurfaceFluxTendency`
* `timeDaily_avg_activeTracerNonLocalTendency_temperatureNonLocalTendency`
* `timeDaily_avg_activeTracerNonLocalTendency_salinityNonLocalTendency`

Some of these we probably want to fix and others we might just want to take the `_FillValue` off of and address in a different PR.

A few variables with dims `nVertLevels` or `nVertLevelsP1` don't have _FillValues:
* `timeDaily_avg_vertNonLocalFlux_vertNonLocalFluxTemp`
* `timeDaily_avg_activeTracersTend_temperatureTend`
* `timeDaily_avg_activeTracersTend_salinityTend`

These last 2 may be on purpose from our discussion of #571 but should be noted so we can fix them later.

wenshanw commented on src/core_ocean/Registry.xml:
```diff
@@ -2608,7 +2682,7 @@
 			<var name="vertNonLocalFluxTemp" array_group="vertNonLocalFlux" units="nondimensional" name_in_code="vertNonLocalFluxTemp"
 				 description="CVMix/KPP: nonlocal boundary layer mixing term for temperature"
-		/>
+			/>
```
I added a tab here because, in other `var_array` environments, the closing parenthesis of `var` aligns with their opening parenthesis. The closing parenthesis for `var_array` is `</var_array>` which is aligned with `<var_array ...>` later in the code.

xylar replied:
Oh, I see.  YOu are right.

wenshanw commented on src/core_ocean/shared/mpas_ocn_diagnostics.F:
```diff
@@ -331,8 +332,8 @@ subroutine ocn_diagnostic_solve(dt, statePool, forcingPool, meshPool, diagnostic
 
       normalVelocity(:,nEdges+1) = -1e34
       layerThickness(:,nCells+1) = -1e34
```
Yes! Sorry, in my first version, these two variables have no `missing values` in the output and no `fill_value` attributes. So I changed them back to -1e34.

Double-checking other variables: Should I replace the following with MPAS_REAL_FILLVAL as well?
```
mode_init/mpas_ocn_init_vertical_grids.F:        localMaxRx1Edge = -1e30_RKIND
mode_init/mpas_ocn_init_vertical_grids.F:        localStretchMin = 1e30_RKIND
mode_init/mpas_ocn_init_vertical_grids.F:        localStretchMax = -1e30_RKIND
mode_init/mpas_ocn_init_interpolation.F:         minDist = 1e34
```

xylar replied:
I think all of these max and min variables are right as they are.  They are very large positive or negative numbers so the first value found will replace them as the new min or max.

xylar commented:
For reference, here is my version of this PR where I broke the addition of fill values into the registry into many more commits to make debugging easier:
https://github.com/xylar/MPAS-Model/tree/ocean/add_fill_values

xylar commented:
For reference, here is a script I used for debugging which fields have valid _FillValues and which have zeros:
``` bash
#!/bin/bash

rm -rf analysis_members/mpaso.hist.am.timeSeriesStatsDaily.0001-01-01.nc
./run.py
ncdump -h analysis_members/mpaso.hist.am.timeSeriesStatsDaily.0001-01-01.nc | grep _FillValue | awk '{print $1;}' | awk -F ":" '{print $1;}' > vars
for var in `cat vars`
do
    echo $var
    ncdump -v $var analysis_members/mpaso.hist.am.timeSeriesStatsDaily.0001-01-01.nc | grep "_," | tail -n 2
    echo " "
    ncdump -v $var analysis_members/mpaso.hist.am.timeSeriesStatsDaily.0001-01-01.nc | grep " 0," | tail -n 2
done
```

wenshanw commented:
> I added all of the `missing_value="FILLVAL"` into the main registry, one `var_struct` at a time. Everything worked for me. I didn't add fill values to fields in the `scratch` `var_struct` because we would never be outputting fields from that. (I certainly hope we aren't!)
> 
> I got the feeling that fill values were added to a lot of variables that aren't in `timeSeriesStatsDaily/Monthly`. That's okay but I wanted to check on that just to understand what your system was for deciding which variables got fill values and which didn't.
> 
> In my testing, the following variables in `timeSeriesStatsDaily` ended up with `_FillValue` attributes but still have `0` as the missing value in the output:
> 
> * `timeDaily_avg_velocityMeridional`
> * `timeDaily_avg_velocityZonal`
> * `timeDaily_avg_tendLayerThickness`
> * `timeDaily_avg_normalGMBolusVelocity`
> * `timeDaily_avg_GMBolusVelocityZonal`
> * `timeDaily_avg_GMBolusVelocityMeridional`
> * `timeDaily_avg_vertDiffTopOfCell`
> * `timeDaily_avg_vertViscTopOfCell`
> * `timeDaily_avg_bulkRichardsonNumber` (?)
> * `timeDaily_avg_tendLayerThickness`
> * `timeDaily_avg_activeTracerVertMixTendency_temperatureVertMixTendency`
> * `timeDaily_avg_activeTracerVertMixTendency_salinityVertMixTendency`
> * `timeDaily_avg_activeTracerSurfaceFluxTendency_temperatureSurfaceFluxTendency`
> * `timeDaily_avg_activeTracerSurfaceFluxTendency_salinitySurfaceFluxTendency`
> * `timeDaily_avg_activeTracerNonLocalTendency_temperatureNonLocalTendency`
> * `timeDaily_avg_activeTracerNonLocalTendency_salinityNonLocalTendency`
> 
> Some of these we probably want to fix and others we might just want to take the `_FillValue` off of and address in a different PR.

OK. I will check if I can replace `zeros` by `MPAS_REAL_FILLVAL` in the code for these variables. If not, remove `_FillValue`.

> 
> A few variables with dims `nVertLevels` or `nVertLevelsP1` don't have _FillValues:
> 
> * `timeDaily_avg_vertNonLocalFlux_vertNonLocalFluxTemp`
> * `timeDaily_avg_activeTracersTend_temperatureTend`
> * `timeDaily_avg_activeTracersTend_salinityTend`
> 
> These last 2 may be on purpose from our discussion of #571 but should be noted so we can fix them later.

Somehow, if I added `fill_value` for the `var_array vertNonLocalFlux` in `Registry.xml`, the case won't run. Same as `var wettingVelocity`. All other variables are fine.

xylar commented:
> Somehow, if I added fill_value for the var_array vertNonLocalFlux in Registry.xml, the case won't run. Same as var wettingVelocity. All other variables are fine.

Okay, that's fine.  Let's make a careful note in the description at the top of this PR of all the variables we want to add `_FillValue` attributes to later but which were not successful in this PR.  We need to change the name accordingly.

wenshanw commented:
> @wenshanw, were you able to run the QU240/daily_output_test` with these changes? I'm seeing an error:
> 
> ```
> Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
> 
> Backtrace for this error:
> #0  0x7f3da239dfcf in ???
> #1  0x564692971d21 in __ocn_time_integration_split_MOD_ocn_time_integrator_split
> 	at /home/xylar/code/mpas-work/model/ocean/test_merge/src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:967
> #2  0x5646929820b3 in __ocn_time_integration_MOD_ocn_timestep
> 	at /home/xylar/code/mpas-work/model/ocean/test_merge/src/core_ocean/mode_forward/mpas_ocn_time_integration.F:111
> ...
> ```

My end is the same as @philipwjones: I got the IEEE warning but the runs finish and the results look fine.

wenshanw commented:
> @wenshanw, since this new PR includes all the changes from #571, can we close that one?

Yes, please.

xylar commented:
@mark-petersen, same as #634, is this something we want to work on soon or move over to E3SM/

mark-petersen commented:
This PR was updated and migrated to the E3SM repo, https://github.com/E3SM-Ocean-Discussion/E3SM/pull/20
